src/me.rs

```diff
@@ -148,113 +148,144 @@ pub fn motion_estimation(
+                     blk_w: usize, blk_h: usize) -> MotionVector
+{
+  let diamond_pattern = [(1i16, 0i16), (0, 1), (-1, 0), (0, -1)];
+  let mut diamond_radius: i16 = 16;
+
+  let mut tmp_plane = Plane::new(blk_w, blk_h, 0, 0, 0, 0);
+
+  let mut center_mv = cmv;
+  let mut center_cost = get_mv_cost(
+    fi, fs, po, ref_frame, bit_depth,
+    pmv, lambda, mvx_min, mvx_max, mvy_min, mvy_max,
+    blk_w, blk_h, center_mv, &mut tmp_plane);
+
+  if center_cost == std::u32::MAX {
+    /* cmv is not a valid motion vector for the current block.
+       We fallback to the (0, 0) center motion vector. */
+    center_mv = MotionVector{row: 0, col: 0};
+    center_cost = get_mv_cost(
+      fi, fs, po, ref_frame, bit_depth,
+      pmv, lambda, mvx_min, mvx_max, mvy_min, mvy_max,
+      blk_w, blk_h, center_mv, &mut tmp_plane);
+  }
+
+  loop {
+
+    let mut best_diamond_cost = std::u32::MAX;
+    let mut best_diamond_mv = MotionVector { row: 0, col: 0 };
 
-      for step in steps {
-        let center_mv_h = best_mv;
-        for i in 0..3 {
-          for j in 0..3 {
-            // Skip the center point that was already tested
-            if i == 1 && j == 1 {
-              continue;
-            }
-
-            let cand_mv = MotionVector {
-              row: center_mv_h.row + step * (i as i16 - 1),
-              col: center_mv_h.col + step * (j as i16 - 1)
-            };
-
-            if (cand_mv.col as isize) < mvx_min || (cand_mv.col as isize) > mvx_max {
-              continue;
-            }
-            if (cand_mv.row as isize) < mvy_min || (cand_mv.row as isize) > mvy_max {
-              continue;
-            }
-
-            {
-              let tmp_slice =
-                &mut tmp_plane.mut_slice(&PlaneOffset { x: 0, y: 0 });
-
-              mode.predict_inter(
-                fi,
-                0,
-                &po,
-                tmp_slice,
-                blk_w,
-                blk_h,
-                [ref_frame, NONE_FRAME],
-                [cand_mv, MotionVector { row: 0, col: 0 }]                
-              );
-            }
-
-            let plane_org = fs.input.planes[0].slice(&po);
-            let plane_ref = tmp_plane.slice(&PlaneOffset { x: 0, y: 0 });
-
-            let sad = get_sad(&plane_org, &plane_ref, blk_h, blk_w, fi.sequence.bit_depth);
-
-            let rate1 = get_mv_rate(cand_mv, pmv[0], fi.allow_high_precision_mv);
-            let rate2 = get_mv_rate(cand_mv, pmv[1], fi.allow_high_precision_mv);
-            let rate = rate1.min(rate2 + 1);
-            let cost = 256 * sad + rate * lambda;
-
-            if cost < lowest_cost {
-              lowest_cost = cost;
-              best_mv = cand_mv;
-            }
-          }
+    for p in diamond_pattern.iter() {
+
+        let cand_mv = MotionVector {
+          row: center_mv.row + diamond_radius * p.0,
+          col: center_mv.col + diamond_radius * p.1
+        };
+
+        let cost = get_mv_cost(
```

Comment:
The 'cost' here includes sad since get_mv_cost() now does sad calculation.
Comment about it would help readers, because traditionally 'mv_cost' means bit cost for the mv only in many video codec codebase and academic papers.

Reply:
Thank you for your comment.
I can also rename the 'get_mv_cost' function and the 'cost' variable if their name are confusing.

Reply:
I'd recommend a very verbose get_mv_and_sad_cost() here! :)

Reply:
As we have discussed in the last daala meeting,
since get_mv_cost() involved lambda, "return 256 * sad + rate * lambda;", the function is essentially return the final RD cost. So, to me it would be best to name the function simply as "get_rd_cost()" I think.
Hence, I'd recommend: let rd_cost = get_rd_cost(...);